### DISCUSSION TITLE
Help with this error: Given indices do not match output variable rank!

### DISCUSSION THREAD

**Discussion by ViniciusBMoreira** [mhm-ufz/mHM#17] [Q&A] [answered]
Hello everybody.
 Congratulations on developing the mHM. I'm trying to apply it to a basin in Brazil, but I find this error during the process.
 I tried to change as little as possible of the mhm file (converted to .txt) [mhm.txt](https://github.com/mhm-ufz/mhm/files/6639531/mhm.txt)

But now i'm stuck in this error, which i couldn't identify the reason. Can you help me?

P.s.: I'm not using any of the optional data in the model. (et.nc, neutrons.nc, Q_bkfl.nc, sm.nc, twsa.nc)

```
----------------------------------------------------------------------
              mHM-UFZ

    MULTISCALE HYDROLOGIC MODEL
           Version: 5.11.2-dev0
           Date:    2021-06-10

Originally by L. Samaniego & R. Kumar
----------------------------------------------------------------------

Openmp not used.
MPI not used.
Start at 11.06.2021 12:23:43.
Working directory: /root/velhas
Using main file mhm_driver.f90 and namelists:
     mhm.nml
     mhm_parameter.nml
     mhm_outputs.nml


Read namelist file: mhm.nml
Read namelist file: mhm_parameter.nml
Read namelist file: mhm_outputs.nml
 INFO: Process (10, neutrons) is deactivated, so output will be suppressed.

domain 1:
resolution Hydrology (domain 1)     = 25.000
resolution Routing (domain 1)       = 25.000

Resolution of routing and hydrological modeling are equal!

Following output will be written:
  NetCDF deflate level: 6
  NetCDF output precision: double
  STATES:
    interceptional storage                          (L1_inter) [mm]
    soil water content in the single layers     (L1_soilMoist) [mm]
    volumetric soil moisture in the single layers              [mm/mm]
    mean volum. soil moisture averaged over all soil layers    [mm/mm]
    waterdepth in reservoir of sealed areas       (L1_sealSTW) [mm]
    waterdepth in reservoir of unsat. soil zone  (L1_unsatSTW) [mm]
  FLUXES:

FINISHED reading config

  Inititalize mRM

    Following output will be written:
    NetCDF deflate level: 6
    NetCDF output precision: double
    FLUXES:
      routed streamflow      (L11_qMod)                [m3 s-1]
      river temperature      (RivTemp)                 [deg C]

    FINISHED reading config


# of domains:                  1

  Input data directories:
  --------------
      DOMAIN                    1
  --------------
    Morphological directory:      test_domain/input/morph/ (found)
    Land cover directory:         test_domain/input/luse/ (found)
    Precipitation directory:      test_domain/input/meteo/pre/ (found)
    Temperature directory:        test_domain/input/meteo/tavg/ (found)
    PET directory:                test_domain/input/meteo/pet/ (found)
    Output directory:             test_domain/output_b1/ (found)


  Read data ...
  Reading data ...
    Reading dem and lcover ...
      Reading dem for domain: 1 ...
      Reading lcover for domain: 1 ...

Tillage layers: the tillage horizons are modelled
                upto mHM layers:          1
    Reading data for domain: 1 ...
      Reading slope ...
      Reading aspect ...
      Reading soil ids ...
      Reading LAI ...
      Reading latitude/logitude ...
    in   114.658 seconds.
  in   158.575 seconds.
***WARNING: Geological unit 2
            is not appearing in study domain.
***WARNING: Geological unit 6
            is not appearing in study domain.
***WARNING: Geological unit 8
            is not appearing in study domain.
***WARNING: Geological unit -9999
            is not appearing in study domain.
    in   158.575 seconds.
  Initialize domains ...
 Given indices do not match output variable rank!
```

**Comment by StephanThober** [comment]
Dear Bryan,

Great to hear that you are using mHM and sorry to see that you run into problems. Please report your issue on https://github.com/mhm-ufz/mhm/discussions and paste your mhm.nml so that we can help you to sort this out.

Kind regards,
Stephan

--------------------------------------------------------------------------------------------
						
Dr. Stephan Thober
Senior scientist
Department Computational Hydrosystems
				
Helmholtz-Zentrum für Umweltforschung GmbH - UFZ
Helmholtz Centre for Environmental Research GmbH - UFZ	
Permoserstraße 15, 04318 Leipzig, Germany
Phone +49 341 235 482324
***@***.***, www.ufz.de/index.php?en=38099

Sitz der Gesellschaft/Registered Office: Leipzig
Registergericht/Registration Office: Amtsgericht Leipzig
Handelsregister Nr./Trade Register Nr.: B 4703
Vorsitzende des Aufsichtsrats/Chairwoman of the Supervisory Board: MinDirig'in Oda Keppler
Wissenschaftlicher Geschäftsführer/Scientific Managing Director: 
Prof. Dr. Georg Teutsch
Administrative Geschäftsführerin/Administrative Managing Director: 
Dr. Sabine König

--------------------------------------------------------------------------------------------

**Reply by ViniciusBMoreira** [reply]
Hey Stephan! ty for your answer. 
Isn't that the discussion channel?

**Comment by StephanThober** [comment]
Hi Bryan,

No this channel is for all sorts of issues that user might run into like the one you stumbled upon.

Best,
Stephan

--------------------------------------------------------------------------------------------
						
Dr. Stephan Thober
Senior scientist
Department Computational Hydrosystems
				
Helmholtz-Zentrum für Umweltforschung GmbH - UFZ
Helmholtz Centre for Environmental Research GmbH - UFZ	
Permoserstraße 15, 04318 Leipzig, Germany
Phone +49 341 235 482324
***@***.***, www.ufz.de/index.php?en=38099

Sitz der Gesellschaft/Registered Office: Leipzig
Registergericht/Registration Office: Amtsgericht Leipzig
Handelsregister Nr./Trade Register Nr.: B 4703
Vorsitzende des Aufsichtsrats/Chairwoman of the Supervisory Board: MinDirig'in Oda Keppler
Wissenschaftlicher Geschäftsführer/Scientific Managing Director: 
Prof. Dr. Georg Teutsch
Administrative Geschäftsführerin/Administrative Managing Director: 
Dr. Sabine König

--------------------------------------------------------------------------------------------

**Reply by ViniciusBMoreira** [reply]
So, i'm in the right place, correct?

**Reply by StephanThober** [reply]
yes, sorry for the confusion.

**Comment by pallavhydro** [comment, marked as answer]
Hi Bryan

I may not be able to point out the issue directly, but I have concerns over the following based on your mhm.nml -

1. You want mhm to run at 25m x 25m grid. Is that correct? You have selected  `iFlag_cordinate_sys` as x, y system (in m) and entered `resolution_Hydrology` and `resolution_Routing` as 25 i.e. 25 m. 

1. In mhm, there are three resolutions viz. L0, L1 (also L11) and L2. L0 - morphology (dem, etc.), L1 - hydrology (model resolution), L11 - routing (use same as L1 if not sure) and L2 - meteorology. Usually we set `L0 < L1/L11 <= L2`. Your L1 is 25 m. What is the L0 dataset resolution that you are using? Perhaps using larger L1 and L11 (say 25000 i.e. 25 km) makes sense. 

2. If you are using L1,L11 = 25m resolution, have you prepared the corresponding latlon file (`latlon.nc`) for your model setup?  

2. Also I suppose you have your basin data placed under `test_domain` folder structure? May I suggest you to perhaps use a separate folder with your basin name to avoid confusion. At the moment, looking at the mhm.nml, it looks as if you are using the test domain data. 

Cheers
Pallav

**Reply by ViniciusBMoreira** [reply]
Hey!

**_You want mhm to run at 25m x 25m grid. Is that correct?_ _What is the L0 dataset resolution that you are using?_**
Yes, all my input data are at 25m resolution. 

**_If you are using L1,L11 = 25m resolution, have you prepared the corresponding latlon file (latlon.nc) for your model setup?_**
Perhaps here is the error. I'm only using one latlon grid file since I'm only using one resolution for the entire model.
I create a new latlon file, for different levels but with the same header. I will give it a try !!
**_Also I suppose you have your basin data placed under test_domain folder structure?_**
yes, I tried renaming my files to the same names as the test basin to avoid errors.

**Reply by ViniciusBMoreira** [reply]
Update:

The error disappeared. Now I'm facing another... apparently for lack of space. I think I will have to change the resolution of level 2.

`Program received signal SIGSEGV: Segmentation fault - invalid memory reference.`

**Reply by pallavhydro** [reply]
Hi Bryan

Great. So for the reference of readers of this discussion blog, the solution was `correct latlon file` in line with the resolutions being used.

Next issue: `Seg. Fault`. Few comments -

1. This is a generic error message (not mhm specific) and the debugging of the issue would need more info/ strategy. Since this is now a different issue, I would say it would make more sense if you open a different discussion topic and we continue there.

2. While we continue this discussion as a new topic, I would suggest you (as already mentioned by @StephanThober ) to again post your latest `mhm.nml` and the `runtime output of the error`.  This will give us some clues already.  But unlike last time, try posting mhm.nml content as code as well. Let's see whether this becomes the better way to share the namelist file or just obstructs readability.

Cheers
Pallav

**Reply by ViniciusBMoreira** [reply]
Thanks a lot Pallav.

First, I will try to solve the problem myself. I believe it has something to do with memory allocation during `cmake`. If I can, I'll post the solution here. Otherwise, I will ask for help. Thanks for the support!

Bryan
